src/main/java/com/ibm/cloud/sdk/core/security/TokenRequestBasedAuthenticator.java

```diff
@@ -202,6 +202,34 @@ public void authenticate(Builder builder) {
    */
   public abstract T requestToken();
 
+  /**
+   * Calls the extending class' requestToken implementation in a synchronized way, returning the existing token data
+   * if it's been made valid since this method's initial call.
+   *
+   * @return the token object
+   */
+  private synchronized T synchronizedRequestToken() {
+    if (this.tokenData != null && this.tokenData.isTokenValid()) {
+      return this.tokenData;
+    }
+
+    return requestToken();
+  }
+
+  /**
+   * Requests a new token. The refresh time for the token is also updated to some point in the future to prevent
+   * other threads from also calling this method while the new token is set.
+   */
+  private synchronized void refreshToken() {
+    // Another thread may have already updated this time. In that case, we can just return.
+    if (this.tokenData.isTokenValid()) {
+      return;
+    }
+
+    this.tokenData.advanceRefreshTime();
```

Comment:
Rather than advancing the refresh time here, I think you should do it within `needsRefresh` whenever it responds `true`.